daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala

```diff
@@ -0,0 +1,349 @@
+  ): Boolean =
+    recordedId match {
+      case None => true
+      case _ => recordedId == replayedId
+    }
+
+  private[this] def keyIsReplayedBy(
+      recorded: Option[KeyWithMaintainers[Value[Cid]]],
+      replayed: Option[KeyWithMaintainers[Value[Cid]]],
+  ): Boolean = {
+    (recorded, replayed) match {
+      case (None, None) => true
+      case (Some(recordedValue), Some(replayedValue)) =>
+        keyIsReplayedBy(recordedValue, replayedValue)
+      case _ => false
+    }
+  }
+
+  private[this] def keyIsReplayedBy(
+      recorded: KeyWithMaintainers[Value[Cid]],
+      replayed: KeyWithMaintainers[Value[Cid]],
+  ): Boolean = {
+    valueIsReplayedBy(recorded.key, replayed.key) && recorded.maintainers == replayed.maintainers
+  }
+
+  private[this] def keys[K](entries: ImmArray[(K, _)]): Iterator[K] =
+    entries.iterator.map(_._1)
+
+  private[this] def values[V](entries: ImmArray[(_, V)]): Iterator[V] =
+    entries.iterator.map(_._2)
+
+  private def valueIsReplayedBy(
+      recorded: Value[Cid],
+      replayed: Value[Cid],
+  ): Boolean = {
+
+    import Value._
+
+    @tailrec
+    def loop(tuples: LazyList[(Value[Cid], Value[Cid])]): Boolean =
+      tuples match {
+        case LazyList.cons(tuple, rest) =>
+          tuple match {
+            case (ValueEnum(recordedTyCon, recordedName), ValueEnum(replayedTyCon, replayedName)) =>
+              tyConIsReplayedBy(recordedTyCon, replayedTyCon) &&
+                recordedName == replayedName &&
+                loop(rest)
+            case (recordedLeaf: ValueCidlessLeaf, replayedLeaf: ValueCidlessLeaf) =>
+              recordedLeaf == replayedLeaf &&
+                loop(rest)
+            case (ValueContractId(recordedValue), ValueContractId(replayedValue)) =>
+              recordedValue === replayedValue &&
+                loop(rest)
+            case (
+                  ValueRecord(recordedTyCon, recordedFields),
+                  ValueRecord(replayedTyCon, replayedFields),
+                ) =>
+              tyConIsReplayedBy(recordedTyCon, replayedTyCon) &&
+                recordedFields.length == replayedFields.length &&
+                (keys(recordedFields) zip keys(replayedFields)).forall {
+                  case (None, _) => true
+                  case (recorded, replayed) => recorded == replayed
+                } &&
+                loop((values(recordedFields) zip values(recordedFields)) ++: rest)
+            case (
+                  ValueVariant(recordedTyCon, recordedVariant, recordedValue),
+                  ValueVariant(replayedTyCon, replayedVariant, replayedValue),
+                ) =>
+              tyConIsReplayedBy(recordedTyCon, replayedTyCon) &&
+              recordedVariant == replayedVariant
+              loop((recordedValue, replayedValue) +: rest)
+            case (ValueList(recordedValues), ValueList(replayedValues)) =>
+              loop((recordedValues.iterator zip replayedValues.iterator) ++: rest)
```

Comment:
don’t we need to check the length of the lists here?

Reply:
:+1